cluster-autoscaler/simulator/basic_cluster_snapshot.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (data *internalBasicSnapshotDataNodeLister) List() ([]*schedulernodeinfo.NodeInfo, error) {
+	nodeInfoList := make([]*schedulernodeinfo.NodeInfo, 0, len(data.nodeInfoMap))
+	for _, v := range data.nodeInfoMap {
+		nodeInfoList = append(nodeInfoList, v)
+	}
+	return nodeInfoList, nil
+}
+
+func (data *internalBasicSnapshotDataNodeLister) HavePodsWithAffinityList() ([]*schedulernodeinfo.NodeInfo, error) {
+	havePodsWithAffinityList := make([]*schedulernodeinfo.NodeInfo, 0, len(data.nodeInfoMap))
+	for _, v := range data.nodeInfoMap {
+		if len(v.PodsWithAffinity()) > 0 {
+			havePodsWithAffinityList = append(havePodsWithAffinityList, v)
+		}
+	}
+	return havePodsWithAffinityList, nil
+}
+
+func (data *internalBasicSnapshotDataNodeLister) Get(nodeName string) (*schedulernodeinfo.NodeInfo, error) {
+	if v, ok := data.nodeInfoMap[nodeName]; ok {
+		return v, nil
+	}
+	return nil, fmt.Errorf("node %s not in snapshot", nodeName)
+}
+
+func (data *internalBasicSnapshotDataPodLister) List(selector labels.Selector) ([]*apiv1.Pod, error) {
+	alwaysTrue := func(p *apiv1.Pod) bool { return true }
+	return data.FilteredList(alwaysTrue, selector)
+}
+
+func (data *internalBasicSnapshotDataPodLister) FilteredList(podFilter schedulerlisters.PodFilter, selector labels.Selector) ([]*apiv1.Pod, error) {
+	pods := make([]*apiv1.Pod, 0)
+	for _, n := range data.nodeInfoMap {
+		for _, pod := range n.Pods() {
+			if podFilter(pod) && selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (data *internalBasicSnapshotData) Pods() schedulerlisters.PodLister {
+	return (*internalBasicSnapshotDataPodLister)(data)
+}
+
+func (data *internalBasicSnapshotData) NodeInfos() schedulerlisters.NodeInfoLister {
+	return (*internalBasicSnapshotDataNodeLister)(data)
+}
+
+// NewEmptySnapshot initializes a Snapshot struct and returns it.
+func newInternalBasicSnapshotData() *internalBasicSnapshotData {
+	return &internalBasicSnapshotData{
+		nodeInfoMap: make(map[string]*schedulernodeinfo.NodeInfo),
+	}
+}
+
+func (data *internalBasicSnapshotData) clone() *internalBasicSnapshotData {
+	clonedNodeInforMap := make(map[string]*schedulernodeinfo.NodeInfo)
+	for k, v := range data.nodeInfoMap {
+		clonedNodeInforMap[k] = v.Clone()
+	}
+	return &internalBasicSnapshotData{
+		nodeInfoMap: clonedNodeInforMap,
+	}
+}
+
+func (data *internalBasicSnapshotData) addNode(node *apiv1.Node) error {
+	if _, found := data.nodeInfoMap[node.Name]; found {
+		return fmt.Errorf("node %s already in snapshot", node.Name)
+	}
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	err := nodeInfo.SetNode(node)
+	if err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo; %v", err)
+	}
+	data.nodeInfoMap[node.Name] = nodeInfo
+	return nil
+}
+
+func (data *internalBasicSnapshotData) removeNode(nodeName string) error {
+	if _, found := data.nodeInfoMap[nodeName]; !found {
+		return fmt.Errorf("node %s not in snapshot", nodeName)
```

Comment:
Also add a const error? It seems like the kind of error you want to ignore fairly often and I prefer:
```
if err := snapshot.RemoveNode(); err != nil && err != ErrNodeNotInSnapshot {
  return err
}
```
to
```
snapshot.RemoveNode()
```

Reply:
Maybe we should just ignore the non-existence of node internally?

Reply:
I'm also ok with not returning error in that case.

Reply:
For now I will not change it. And implement constant errors as a followup. I kinda tend to be on more strict error checking side. Though I feel like with explicit error checking (without exceptions) it renders code harder to rid a bit.